Incorrect span reported for errors originating from proc macros with `cargo_json` analyzer

The reported span is from the file where the macro is defined, instead of the correct span of the file where the error/warning produced.

You can reproduce a warning with the following code:

```rust
// main.rs
use std::fmt::Write;

fn main() {
    let mut s = String::new();
    writeln!(&mut s, "hello"); // this line produces a warning that the `Result` must be used
}
```

([Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=4bfd8c8beab4551088e376592d853e1f))

When running this through `cargo check --message-format json`, the following output is produced:

<details>
<summary>json output</summary>

```json
{
  "reason": "compiler-message",
  "package_id": "path+file:///local/home/<redacted>/tmp/empty#0.1.0",
  "manifest_path": "/local/home/<redacted>/tmp/empty/Cargo.toml",
  "target": {
    "kind": [
      "bin"
    ],
    "crate_types": [
      "bin"
    ],
    "name": "empty",
    "src_path": "/local/home/<redacted>/tmp/empty/src/main.rs",
    "edition": "2021",
    "doc": true,
    "doctest": false,
    "test": true
  },
  "message": {
    "rendered": "warning: unused `Result` that must be used\n --> src/main.rs:5:5\n  |\n5 |     writeln!(&mut s, \"hello\");\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n  = note: this `Result` may be an `Err` variant, which should be handled\n  = note: `#[warn(unused_must_use)]` on by default\n  = note: this warning originates in the macro `writeln` (in Nightly builds, run with -Z macro-backtrace for more info)\n\n",
    "$message_type": "diagnostic",
    "children": [
      {
        "children": [],
        "code": null,
        "level": "note",
        "message": "this `Result` may be an `Err` variant, which should be handled",
        "rendered": null,
        "spans": []
      },
      {
        "children": [],
        "code": null,
        "level": "note",
        "message": "`#[warn(unused_must_use)]` on by default",
        "rendered": null,
        "spans": []
      }
    ],
    "code": {
      "code": "unused_must_use",
      "explanation": null
    },
    "level": "warning",
    "message": "unused `Result` that must be used",
    "spans": [
      {
        "byte_end": 24374,
        "byte_start": 24325,
        "column_end": 58,
        "column_start": 9,
        "expansion": {
          "def_site_span": {
            "byte_end": 24211,
            "byte_start": 24191,
            "column_end": 21,
            "column_start": 1,
            "expansion": null,
            "file_name": "/local/home/<redacted>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/macros/mod.rs",
            "is_primary": false,
            "label": null,
            "line_end": 683,
            "line_start": 683,
            "suggested_replacement": null,
            "suggestion_applicability": null,
            "text": [
              {
                "highlight_end": 21,
                "highlight_start": 1,
                "text": "macro_rules! writeln {"
              }
            ]
          },
          "macro_decl_name": "writeln!",
          "span": {
            "byte_end": 94,
            "byte_start": 69,
            "column_end": 30,
            "column_start": 5,
            "expansion": null,
            "file_name": "src/main.rs",
            "is_primary": false,
            "label": null,
            "line_end": 5,
            "line_start": 5,
            "suggested_replacement": null,
            "suggestion_applicability": null,
            "text": [
              {
                "highlight_end": 30,
                "highlight_start": 5,
                "text": "    writeln!(&mut s, \"hello\");"
              }
            ]
          }
        },
        "file_name": "/local/home/<redacted>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/macros/mod.rs",
        "is_primary": true,
        "label": null,
        "line_end": 688,
        "line_start": 688,
        "suggested_replacement": null,
        "suggestion_applicability": null,
        "text": [
          {
            "highlight_end": 58,
            "highlight_start": 9,
            "text": "        $dst.write_fmt($crate::format_args_nl!($($arg)*))"
          }
        ]
      }
    ]
  }
}
```

</details>

The span is reported in `.message.spans[0].expansion.span`, you can try this for yourself with `jq`:

```bash
$ cargo c --message-format json 2>/dev/null | jq 'select(.reason == "compiler-message") | .message.spans[0].expansion.span'
{
  "byte_end": 94,
  "byte_start": 69,
  "column_end": 30,
  "column_start": 5,
  "expansion": null,
  "file_name": "src/main.rs",
  "is_primary": false,
  "label": null,
  "line_end": 5,
  "line_start": 5,
  "suggested_replacement": null,
  "suggestion_applicability": null,
  "text": [
    {
      "highlight_end": 30,
      "highlight_start": 5,
      "text": "    writeln!(&mut s, \"hello\");"
    }
  ]
}
```

This, however, is different from where the span in a "traditional" diagnostic is located (`.message.children[0].spans[0]`). Therefore, I assume that the wrong file is picked up incorrectly due to the other format.

<details><summary>"Traditional" diagnostic json</summary>
<p>

```json
{
  "reason": "compiler-message",
  "package_id": "path+file:///local/home/<redacted>/tmp/empty#0.1.0",
  "manifest_path": "/local/home/<redacted>/tmp/empty/Cargo.toml",
  "target": {
    "kind": [
      "bin"
    ],
    "crate_types": [
      "bin"
    ],
    "name": "empty",
    "src_path": "/local/home/<redacted>/tmp/empty/src/main.rs",
    "edition": "2021",
    "doc": true,
    "doctest": false,
    "test": true
  },
  "message": {
    "rendered": "error[E0425]: cannot find value `err` in this scope\n   --> src/main.rs:5:5\n    |\n5   |     err;\n    |     ^^^ help: a tuple variant with a similar name exists: `Err`\n    |\n   ::: /local/home/<redacted>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:537:5\n    |\n537 |     Err(#[stable(feature = \"rust1\", since = \"1.0.0\")] E),\n    |     --- similarly named tuple variant `Err` defined here\n\n",
    "$message_type": "diagnostic",
    "children": [
      {
        "children": [],
        "code": null,
        "level": "help",
        "message": "a tuple variant with a similar name exists",
        "rendered": null,
        "spans": [
          {
            "byte_end": 78,
            "byte_start": 75,
            "column_end": 8,
            "column_start": 5,
            "expansion": null,
            "file_name": "src/main.rs",
            "is_primary": true,
            "label": null,
            "line_end": 5,
            "line_start": 5,
            "suggested_replacement": "Err",
            "suggestion_applicability": "MaybeIncorrect",
            "text": [
              {
                "highlight_end": 8,
                "highlight_start": 5,
                "text": "    err;"
              }
            ]
          }
        ]
      }
    ],
    "code": {
      "code": "E0425",
      "explanation": "An unresolved name was used.\n\nErroneous code examples:\n\n```compile_fail,E0425\nsomething_that_doesnt_exist::foo;\n// error: unresolved name `something_that_doesnt_exist::foo`\n\n// or:\n\ntrait Foo {\n    fn bar() {\n        Self; // error: unresolved name `Self`\n    }\n}\n\n// or:\n\nlet x = unknown_variable;  // error: unresolved name `unknown_variable`\n```\n\nPlease verify that the name wasn't misspelled and ensure that the\nidentifier being referred to is valid for the given situation. Example:\n\n```\nenum something_that_does_exist {\n    Foo,\n}\n```\n\nOr:\n\n```\nmod something_that_does_exist {\n    pub static foo : i32 = 0i32;\n}\n\nsomething_that_does_exist::foo; // ok!\n```\n\nOr:\n\n```\nlet unknown_variable = 12u32;\nlet x = unknown_variable; // ok!\n```\n\nIf the item is not defined in the current module, it must be imported using a\n`use` statement, like so:\n\n```\n# mod foo { pub fn bar() {} }\n# fn main() {\nuse foo::bar;\nbar();\n# }\n```\n\nIf the item you are importing is not defined in some super-module of the\ncurrent module, then it must also be declared as public (e.g., `pub fn`).\n"
    },
    "level": "error",
    "message": "cannot find value `err` in this scope",
    "spans": [
      {
        "byte_end": 19919,
        "byte_start": 19916,
        "column_end": 8,
        "column_start": 5,
        "expansion": null,
        "file_name": "/local/home/<redacted>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs",
        "is_primary": false,
        "label": "similarly named tuple variant `Err` defined here",
        "line_end": 537,
        "line_start": 537,
        "suggested_replacement": null,
        "suggestion_applicability": null,
        "text": [
          {
            "highlight_end": 8,
            "highlight_start": 5,
            "text": "    Err(#[stable(feature = \"rust1\", since = \"1.0.0\")] E),"
          }
        ]
      },
      {
        "byte_end": 78,
        "byte_start": 75,
        "column_end": 8,
        "column_start": 5,
        "expansion": null,
        "file_name": "src/main.rs",
        "is_primary": true,
        "label": null,
        "line_end": 5,
        "line_start": 5,
        "suggested_replacement": null,
        "suggestion_applicability": null,
        "text": [
          {
            "highlight_end": 8,
            "highlight_start": 5,
            "text": "    err;"
          }
        ]
      }
    ]
  }
}
```

</p>
</details> 


---

I'm using the following bacon job:

```toml
[jobs.bacon-ls]
command = [ "cargo", "clippy", "--workspace", "--all-targets", "--all-features", "--message-format", "json-diagnostic-rendered-ansi" ]
analyzer = "cargo_json"
need_stdout = true

[exports.cargo-json-spans]
auto = true
exporter = "analyzer"
line_format = "{diagnostic.level}|:|{span.file_name}|:|{span.line_start}|:|{span.line_end}|:|{span.column_start}|:|{span.column_end}|:|{diagnostic.message}|:|{diagnostic.rendered}|:|{span.suggested_replacement}"
path = ".bacon-locations"
```

--- a/src/analysis/cargo_json/cargo_json_export.rs
+++ b/src/analysis/cargo_json/cargo_json_export.rs
@@ -38,7 +38,14 @@ impl CargoJsonExport {
         diagnostic: &Diagnostic,
     ) {
         for span in &diagnostic.spans {
-            let data = OnSpanData { diagnostic, span };
+            let data = {
+                // This is a diagnostic that originates from a proc-macro.
+                if let Some(expansion) = &span.expansion {
+                    OnSpanData { diagnostic, span: &expansion.span }
+                } else {
+                    OnSpanData { diagnostic, span }
+                }
+            };
             let line = self.line_template.render(&data);
             if !line.is_empty() {
                 self.export.push_str(&line);

